### PR TITLE
Allow HttpSys zero-byte reads #41305

### PR DESCRIPTION
**Pull request by Tratcher**
Contributes to #41305

When passed a zero-byte buffer Http.Sys returns ERROR_MORE_DATA. We need to handle this specially. I also had to change the argument validation.

Perf impact:
- zero-byte reads to Http.Sys always complete async, even when data is available and a normal read could complete sync. This adds latency to the request.
- Setup: Upload a 20mb request body using a 16k read buffer.
  - No zero-byte read: 65ms
  - With zero-byte reads: 105ms.

### CODE REVIEW

File: src/Servers/HttpSys/src/RequestProcessing/RequestStream.cs [hunk 1]
```diff
@@ -104,7 +104,7 @@ private static void ValidateReadBuffer(byte[] buffer, int offset, int size)
         {
             throw new ArgumentOutOfRangeException(nameof(offset), offset, string.Empty);
         }
-        if (size <= 0 || size > buffer.Length - offset)
+        if (size < 0 || size > buffer.Length - offset)
```
Comment (gfoidl) [inline src/Servers/HttpSys/src/RequestProcessing/RequestStream.cs:107]:
```suggestion
        if ((uint)size > (uint)(buffer.Length - offset))
```
Saves one comparison. Same could be done in L103 (GH doesn't allow me to comment there).

File: src/Servers/HttpSys/test/FunctionalTests/RequestBodyTests.cs [hunk 2]
```diff
@@ -57,6 +79,29 @@ public async Task RequestBody_ReadAsync_Success()
...
 
+    [ConditionalFact]
+    public async Task RequestBody_Read0ByteAsync_Success()
+    {
+        string address;
+        using (Utilities.CreateHttpServer(out address, async httpContext =>
+        {
+            Assert.True(httpContext.Request.CanHaveBody());
+            byte[] input = new byte[100];
+            await Task.Delay(1000);
```
Comment (JamesNK) [inline src/Servers/HttpSys/test/FunctionalTests/RequestBodyTests.cs:90]: What is the Task.Delay for?

File: src/Servers/HttpSys/test/FunctionalTests/RequestBodyTests.cs [hunk 3]
```diff
@@ -57,6 +79,29 @@ public async Task RequestBody_ReadAsync_Success()
...
+            int read = await httpContext.Request.Body.ReadAsync(input, 0, 1);
```
Comment (JamesNK) [inline src/Servers/HttpSys/test/FunctionalTests/RequestBodyTests.cs:91]: This initial read in the async test isn't in the sync test. Is that intentional?

File: src/Servers/HttpSys/test/FunctionalTests/RequestBodyTests.cs (hunk 2 above)
Comment (Tratcher) [reply src/Servers/HttpSys/test/FunctionalTests/RequestBodyTests.cs:90]: Woops, leftover debugging code 😁.

**Review by JamesNK — APPROVED** [review]

**Pull request merged**